smol/moca/ensemble/ensemble.py

```diff
@@ -0,0 +1,391 @@
+            self.__delete__(obj)
+            return
+
+        for spec, count in Counter(map(get_species, value.keys())).items():
+            if count > 1:
+                raise ValueError(
+                    f"{count} values of the chemical potential for the same "
+                    f"species {spec} were provided.\n Make sure the dictionary "
+                    "you are using has only string keys or only Species "
+                    "objects as keys."
+                )
+        value = {
+            get_species(k): v for k, v in value.items() if get_species(k) in obj.species
+        }
+        if set(value.keys()) != set(obj.species):
+            raise ValueError(
+                "Chemical potentials given are missing species. "
+                "Values must be given for each of the following:"
+                f" {obj.species}"
+            )
+
+        # if first instantiation concatenate the natural parameter
+        if not hasattr(obj, self.private_name):
+            obj.natural_parameters = np.append(
+                obj.natural_parameters, self.natural_parameter
+            )
+        setattr(
+            obj,
+            self.private_name,
+            {"value": ChemicalPotential(value), "table": self._build_table(obj, value)},
+        )
+        # update the ensemble dictionary and _boundaries list
+        if hasattr(obj, "thermo_boundaries"):
+            obj.thermo_boundaries.update({self.public_name: value})
+        else:
+            setattr(obj, "thermo_boundaries", {self.public_name: value})
+
+    def __delete__(self, obj):
+        """Delete the boundary condition."""
+        if hasattr(obj, self.private_name):
+            del obj.__dict__[self.private_name]
+        if (
+            hasattr(obj, "thermo_boundaries")
+            and self.public_name in obj.thermo_boundaries
+        ):
+            del obj.thermo_boundaries[self.public_name]
+        if obj.num_energy_coefs < len(obj.natural_parameters):
+            obj.natural_parameters = obj.natural_parameters[:-1]  # remove last entry
+
+    @staticmethod
+    def _build_table(obj, value):
+        """Set the chemical potentials and update table."""
+        num_cols = max(max(sl.encoding) for sl in obj.sublattices) + 1
+
+        # Sublattice can only be initialized as default, or splitted from default.
+        table = np.zeros((obj.num_sites, num_cols))
+        for sublatt in obj.active_sublattices:
+            ordered_pots = [value[sp] for sp in sublatt.site_space]
+            table[sublatt.sites[:, None], sublatt.encoding] = ordered_pots
+        return table
+
+
+class Ensemble(MSONable):
+    """Thermodynamic ensemble class.
+
+    Attributes:
+        thermo_boundaries (dict):
+            Dictionary with corresponding thermodynamic boundaries, i.e.
+            chemical potentials or fugacity fractions. This is kept only for
+            descriptive purposes.
+    """
+
+    chemical_potentials = ChemicalPotentialManager()
```

Comment:
Good question. It depends how you copy the instances. I did not implement a copy instance, but my guess is if you use `copy` then things will be shared (as in if you change the chemical potentials in a copy then it will also change it in the original). If you use `deepcopy` I think that wont be the case.

I have also seen in pymatgen, and I think I have it for some classes, a copy method that simply uses the `as_dict` method. Basically this,
```python
def copy(self):
    return Ensemble.from_dict(self.as_dict())
```
This will give a deep copy where the two copies will have all independent attributes. But have not thought over all the repercussions. Though if you want a copy method we can simply add this to the class.

Reply:
Thanks! Learned a lot about this type of descriptor grammar.